pkg/nitrite/nitrite.go

```diff
@@ -0,0 +1,122 @@
+// Package nitrite implements attestation verification for AWS Nitro Enclaves.
+package nitrite
+
+import (
+	"crypto/x509"
+	"time"
+
+	"github.com/blocky/nitrite/pkg/attestation"
+	"github.com/blocky/nitrite/pkg/cose"
+)
+
+// Result is a successful verification result of an attestation payload.
+type Result struct {
+	// Document contains the attestation document.
+	Document *attestation.Document `json:"document,omitempty"`
+
+	// Certificates contains all of the certificates except the root.
+	Certificates []*x509.Certificate `json:"certificates,omitempty"`
+
+	// Protected section from the COSE Sign1 payload.
+	Protected []byte `json:"protected,omitempty"`
+	// Unprotected section from the COSE Sign1 payload.
+	Unprotected []byte `json:"unprotected,omitempty"`
+	// Payload section from the COSE Sign1 payload.
+	Payload []byte `json:"payload,omitempty"`
+	// Signature section from the COSE Sign1 payload.
+	Signature []byte `json:"signature,omitempty"`
+
+	// SignatureOK designates if the signature was OK (but certificate could be
+	// invalid, not trusted, expired, etc.)
+	SignatureOK bool `json:"signature_ok"`
+
+	// COSESign1 contains the COSE Signature Structure which was used to
+	// calculate the `Signature`.
+	COSESign1 []byte `json:"cose_sign1,omitempty"`
+}
+
+// VerifyOptions specifies the options for verifying the attestation payload.
+// If `Roots` is nil, the `DefaultCARoot` is used. If `CurrentTime` is 0,
+// `time.Now()` will be used. It is a strong recommendation you explicitly
+// supply this value.
+type VerifyOptions struct {
+	Roots               *x509.CertPool
+	CurrentTime         time.Time
+	AllowSelfSignedCert bool
+}
+
+// Verify verifies the attestation payload from `data` with the provided
+// verification options. If the options specify `Roots` as `nil`, the
+// `DefaultCARoot` will be used. If you do not specify `CurrentTime`,
+// `time.Now()` will be used. It is strongly recommended you specifically
+// supply the time.  If the returned error is non-nil, it is either one of the
+// `Err` codes specified in this package, or is an error from the `crypto/x509`
+// package. Revocation checks are NOT performed and you should check for
+// revoked certificates by looking at the `Certificates` field in the `Result`.
+// Result will be non-null if and only if either of these are true: certificate
+// verification has passed, certificate verification has failed (expired, not
+// trusted, etc.), signature is OK or signature is not OK. If either signature
+// is not OK or certificate can't be verified, both Result and error will be
+// set! You can use the SignatureOK field from the result to distinguish
+// errors.
+func Verify(data []byte, options VerifyOptions) (*Result, error) {
+	cosePayload, err := cose.ExtractCosePayload(data)
```

Comment:
I think you can simplify this quite a bit. To get the `cosePayload` all you need to do is call `cbor.Unmarshal(data, &cose)`. From there you can call `cose.Verify()`, which will got through all the extraction and verification processes below. The idea is that a `COSEPayload` knows how to verify itself and can ask each of its components to verify itself. The `Verify` function here does not really need to know what makes up a `COSEPayload` and how to verify its components.